### PR TITLE
Stronger address book typing

### PR DESCRIPTION
**Pull request by connorwstein**
## Motivation
Want to be maximally defensive for the address book to prevent corruption. 

## Solution
- Type contract and version
- Let products compose type and version at deploy/state gen time to avoid a combinatorial number of strings

### CODE REVIEW

File: integration-tests/deployment/address_book.go [hunk 1]
```diff
@@ -1,25 +1,106 @@
...
-func (m *AddressBookMap) Save(chainSelector uint64, address string, typeAndVersion string) error {
+func (m *AddressBookMap) Save(chainSelector uint64, address string, typeAndVersion TypeAndVersion) error {
+	_, exists := chainsel.ChainBySelector(chainSelector)
+	if !exists {
+		return fmt.Errorf("chain selector %d not found", chainSelector)
+	}
+	if address == "" {
+		return fmt.Errorf("address cannot be empty")
+	}
+	if common.IsHexAddress(address) {
```
Comment (AnieeG) [inline integration-tests/deployment/address_book.go:98]: is zero address allowed?

File: integration-tests/deployment/address_book_test.go [hunk 2]
```diff
@@ -3,69 +3,78 @@ package deployment
...
-		3: {
-			"0x3": "OnRamp 1.0.0",
+		chainsel.TEST_90000003.Selector: {
+			addr3: onRamp100,
 		},
 	})
 
 	// Merge to an existing chain.
-	require.NoError(t, ab2.Save(2, "0x3", "OffRamp 1.0.0"))
+	require.NoError(t, ab2.Save(chainsel.TEST_90000002.Selector, addr3, offRamp100))
```
Comment (AnieeG) [inline integration-tests/deployment/address_book_test.go:92]: is it a use case ( if yes, good to include in test) -
destination and source addressbookMap - both have same selector but diff address and typeVersion? For example -
source - 
```
map[uint64]map[string]TypeAndVersion{
		chainsel.TEST_90000003.Selector: {
			addr4: onRamp110,
		}
}
```
and destination -
```
map[uint64]map[string]TypeAndVersion{
		chainsel.TEST_90000003.Selector: {
			addr3: onRamp100,
		}
}	
```	
What are we expecting out of that merge?

File: integration-tests/deployment/ccip/deploy.go [hunk 3]
```diff
@@ -29,26 +30,21 @@ import (
...
-	// 1.5
-	TokenAdminRegistry_1_5_0 = "TokenAdminRegistry 1.5.0-dev"
-	// 1.6
-	CapabilitiesRegistry_1_0_0 = "CapabilitiesRegistry 1.0.0"
-	CCIPConfig_1_6_0           = "CCIPConfig 1.6.0-dev"
-	EVM2EVMMultiOnRamp_1_6_0   = "EVM2EVMMultiOnRamp 1.6.0-dev"
-	EVM2EVMMultiOffRamp_1_6_0  = "EVM2EVMMultiOffRamp 1.6.0-dev"
-	NonceManager_1_6_0         = "NonceManager 1.6.0-dev"
-	PriceRegistry_1_6_0        = "PriceRegistry 1.6.0-dev"
+	MockARM              deployment.ContractType = "MockARM"
```
Comment (AnieeG) [inline integration-tests/deployment/ccip/deploy.go:33]: ARM or RMN?

File: integration-tests/deployment/address_book_test.go (hunk 2 above)
Comment (connorwstein) [reply integration-tests/deployment/address_book_test.go:92]: as long as the (chain, address) is unique, it should be allowed. So the merge of that would include both

File: integration-tests/deployment/address_book.go (hunk 1 above)
Comment (connorwstein) [reply integration-tests/deployment/address_book.go:98]: can't think of a use case. Will disallow

File: integration-tests/deployment/ccip/deploy.go (hunk 3 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/deploy.go:33]: this one is still the old MockARM wrapper

Comment (makramkd) [reply integration-tests/deployment/ccip/deploy.go:33]: @RensR has a rename PR out for this

**Review by makramkd — APPROVED** [review]

**Pull request merged**